src/main.cairo

```diff
@@ -1,32 +1,182 @@
+}
+
+@constructor
+func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    naming_address: felt, 
+    pricing_address: felt
+) {
+    naming_contract.write(naming_address);
+    pricing_contract.write(pricing_address);
+    return ();
+}
+
+// @notice Get the status of renewals for a domain
+// @param domain Domain to get status of
+// @param user User to get status of
+@view
+func is_renewing{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt, 
+    renewer: felt
+) -> (res: felt) {
+    let (res) = renews.read(renewer, domain);
+    return (res,);
+}
+
+// @notice Get the status of an upgrade vote
+// @param upgrade Upgrade to get status of
+// @param user User to get status of
+@view
+func has_voted_upgrade{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    user: felt,
+    upgrade: felt, 
+) -> (res: felt) {
+    let (res) = voted_upgrade.read(user, upgrade);
+    return (res,);
 }
 
+//
+// Externals
+//
+
+// @notice Activate or deactivate renewals for a domain
+// @param domain Domain to activate or deactivate renewals for
+// @dev callable by owner of domain
+@external
+func toggle_renewals{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt,
+) {
+    alloc_locals;
+    let (caller) = get_caller_address();
+    let (contract) = naming_contract.read();
+
+    let (prev_renew) = renews.read(caller, domain);
+    renews.write(caller, domain, 1 - prev_renew);
+
+    ToggledRenewal.emit(domain, caller, 1 - prev_renew);
+
+    return ();
+} 
+
+// @notice Renew a domain
+// @param domain Domain to renew
+// @dev callable by anyone, but will only renew if renewals are activated for domain
 @external
-func increase_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
-    amount: felt
+func renew{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    root_domain: felt, renewer: felt,
 ) {
-    with_attr error_message("Amount must be positive. Got: {amount}.") {
-        assert_nn(amount);
+    alloc_locals;
+    let (naming) = naming_contract.read(); 
+    let (can_renew) = renews.read(renewer, root_domain);
+
+    with_attr error_message("Renewer has not activated renewals for this domain") {
+        assert_not_zero(can_renew);
+    }
+
+    let (expiry) = Naming.domain_to_expiry(naming, 1, cast(new(root_domain), felt*));
+    let (block_timestamp) = get_block_timestamp();
+    with_attr error_message("Domain is not set to expire within a month") {
+        assert_le(expiry, block_timestamp + 86400 * 30);
+    }
+
+    let (pricing) = pricing_contract.read();
+    let (erc20, renewal_price) = Pricing.compute_renew_price(pricing, root_domain, 365);
+
+    let (contract) = get_contract_address();
+    with_attr error_message("Error transferring tokens from renewer to contract") {
+        IERC20.transferFrom(erc20, renewer, contract, renewal_price);
```

Comment:
Shouldn't we send money to the naming contract address and not this one ?

Reply:
We need to send it first to this contract so the contract has the funds to renew the domains when it calls the `renew` function from the naming contract